lib/rust/lexer/definition/src/token.rs

```diff
@@ -0,0 +1,572 @@
+        let shape  = Shape::Blank;
+        let length = 1;
+        Token{shape,length,offset}
+    }
+
+    /// Construct a token representing an operator.
+    pub fn Operator(name:impl Into<String>, offset:usize) -> Token {
+        let str    = name.into();
+        let length = str.chars().count();
+        let shape  = Shape::Operator(str);
+        Token{shape,length,offset}
+    }
+
+    /// Construct a token representing a modifier operator.
+    pub fn Modifier(name:impl Into<String>, offset:usize) -> Token {
+        let str = name.into();
+        let length = str.chars().count() + 1;
+        let shape = Shape::Modifier(str);
+        Token{shape,length,offset}
+    }
+
+    /// Construct a token representing a number literal.
+    pub fn Number(base:impl Into<String>, num:impl Into<String>, offset:usize) -> Token {
+        let str      = num.into();
+        let base_str = base.into();
+        let length   = if base_str.is_empty() {
+            str.chars().count()
+        } else {
+            base_str.chars().count() + 1 + str.chars().count()
+        };
+        let shape = Shape::Number{base:base_str,number:str};
+        Token{shape,length,offset}
+    }
+
+    /// Construct a token representing a dangling number base.
+    pub fn DanglingBase(base:impl Into<String>, offset:usize) -> Token {
+        let base_str = base.into();
+        let length   = base_str.chars().count() + 1;
+        let shape    = Shape::DanglingBase(base_str);
+        Token{shape,length,offset}
+    }
+
+    /// Construct a token representing a text literal.
+    pub fn Text(text:impl Into<String>, offset:usize) -> Token {
+        let str    = text.into();
+        let length = str.chars().count();
+        let shape  = Shape::Text(str);
+        Token{shape,length,offset}
+    }
+
+    /// Construct a token representing a line of tokens.
+    pub fn Line(tokens:Vec<Token>, offset:usize, trailing_line_ending:LineEnding) -> Token {
+        let line_ending_len = trailing_line_ending.size();
+        let length          = tokens.iter().fold(line_ending_len,|l,r| l + r.offset + r.length);
+        let shape           = Shape::Line{tokens,trailing_line_ending};
+        Token{shape,length,offset}
+    }
+
+    /// Construct a token representing a blank line.
+    ///
+    /// The `offset` for blank lines is from the leftmost column, not from the parent block's
+    /// indentation.
+    pub fn BlankLine(offset:usize, trailing_line_ending:LineEnding) -> Token {
+        let length = trailing_line_ending.size();
+        let shape  = Shape::BlankLine(trailing_line_ending);
+        Token{shape,length,offset}
+    }
+
+    /// Construct a token representing a block.
+    pub fn Block
+    ( block_type  : BlockType
+    , indent      : usize
+    , lines       : Vec<Token>
+    , offset      : usize
+    ) -> Token {
+        let length = lines.iter().map(|line| {
+            let line_length = line.length;
+            let line_offset = line.offset;
```

Comment:
I don't really find having these two identifiers helpful. 
Also, we could somewhat shorten this and in lambda argument decompose the token: 
```
lines.iter().map(|Token{shape,offset,length}| { … 
```
but if you fiknd implicit `line.sth` more readable, it's all good